Codes/Visualization/ConstructOrderBookVisualization_20200214.R

```diff
@@ -45,41 +61,95 @@ mydata$clocktime<-format(as.POSIXct((mydata$time),origin="1970-01-01",tz="UTC"),
-#discard orders that have been completely executed/cancelled by timestamp
-i<-match(subt$orderID,totalect$orderID)
-subt$execanc<-totalect$size[i]; rm(i) #this is the amount of the submitted order that has been executed or cancelled as of timestamp
-subt$execanc<-replace(subt$execanc,is.na(subt$execanc)==1,0) #replace missing with 0
-subt$remainder<-subt$size-subt$execanc #this is the amount of the order remaining as of the timestamp
-  #q<-which(subt$remainder<0) #sanity check
+execanct<-execanc[which(execanc$orderID%in%subt$orderID&execanc$time<testtime&execanc$time>=oldtesttime),]
+      #sum across executions and cancellations (may be several partial or combo)
+      if (nrow(execanct)!=0){
+      totalect<-aggregate(execanct$size,by=list(Category=execanct$orderID),FUN=sum); colnames(totalect)<-c("orderID","size")
+      #discard orders that have been completely executed/cancelled by timestamp
+      i<-match(subt$orderID,totalect$orderID)
+      subt$execanc<-totalect$size[i]; rm(i) #this is the amount of the submitted order that has been executed or cancelled as of timestamp
+      subt$execanc<-replace(subt$execanc,is.na(subt$execanc)==1,0) #replace missing with 0
+      subt$remainder<-subt$size-subt$execanc #this is the amount of the order remaining as of the timestamp
+      if (length(q<-which(subt$remainder<0))!=0){stop("Negative remainders")} #sanity check
+      rm(execanct,totalect)
+      }
+      else {subt$remainder<-subt$size}
 subt<-subt[subt$remainder!=0,] #these should represent the orders that have not yet been execulated
-rm(execanct,totalect)
+#discard buy orders that are higher than the best bid and sell orders that are lower than the best ask
+subt<-subt[which((subt$direction==1&subt$price<=orbk$bidprice1)|(subt$direction==-1&subt$price>=orbk$askprice1)),]
 
 #aggregate across MPID type, price and direction
 subt$temp<-paste(subt$mpid,subt$price,subt$direction,sep=";")
-temp<-aggregate(subt$size,by=list(Category=subt$temp),FUN=sum)
-
-s<-t(matrix(unlist(strsplit(temp$Category,split=";")),3,nrow(temp)))
-temp$Category<-NULL
-temp$mpid<-s[,1]
-temp$price<-as.numeric(s[,2])
-temp$direction<-as.integer(s[,3])
+suba<-aggregate(subt$size,by=list(Category=subt$temp),FUN=sum)
+s<-t(matrix(unlist(strsplit(suba$Category,split=";")),3,nrow(suba)))
+suba$Category<-NULL
+suba$mpid<-s[,1]
+suba$price<-as.numeric(s[,2])
+suba$direction<-as.integer(s[,3])
 rm(s)
+suba<-suba[order(suba$price),]
 
+#only keep x best buy and sell orders
+#x<-50
+#buy<-suba[suba$direction==1,]; buy<-buy[order(buy$price,decreasing=T),]
+#buy<-buy[1:x,]; buy<-buy[order(buy$price),]
+#sell<-suba[suba$direction==-1,]; sell<-sell[order(sell$price),]
+#sell<-sell[1:x,]; sell<-sell[order(sell$price),]
+#suba<-rbind(buy,sell)
 
-##subt[which(subt$orderID=="1059989"),]
+#subt[which(subt$orderID=="1059989"),]
 #totalect[which(totalect$orderID=="1059989"),]
 
-
+suba$interval<-y
+colnames(suba)<-cols
+subplot<-rbind(subplot,suba)
+subold<-subt[,1:48]
```

Comment:
Does this mean that there are 49 level on both side in your plot?

Reply:
Here, I am selecting columns #1-48. This is simply because, in the next loop, I am going the stack the dataframe "subold" on top of the dataframe "subt", and I need the columns to match. Therefore, I am simply getting rid of the extra columns (columns #49+) that were added during the loop.

Reply:
Cool. get it